### PR TITLE
fix(derived_code_mappings): Do not keep hitting Github when API rate limit reached

### PR DESCRIPTION
**Pull request by armenzg**

### CODE REVIEW

File: src/sentry/integrations/github/client.py [hunk 1]
```diff
@@ -118,7 +118,8 @@ def get_tree(self, repo_full_name: str, tree_sha: str) -> JSONData:
             elif msg == "Not Found":
                 logger.warning(f"The Github App does not have access to {repo_full_name}.")
             else:
-                logger.exception("An unknown error has ocurred.")
```
Comment (armenzg) [inline src/sentry/integrations/github/client.py:121]: We would simply log the issue and continue with all other repositories, thus, hammering Github.

Here's [an event](https://sentry.io/organizations/sentry/discover/sentry:e24949f714a549078b15b600b9e23f4e/?field=message&field=organization.slug&interval=1h&name=ApiError%3A+%7B%22message%22%3A%22Not+Found%22%2C%22documentation_url%22%3A%22https%3A%2F%2Fdocs.github.com%2Frest%2Freference%2Fapps%23create-an...&project=1&query=message%3A%22An+unknown+error+has+ocurred.+%7B%5C%22message%5C%22%3A%5C%22API+rate+limit+exceeded+for+installation+ID+865116.%5C%22%2C%5C%22documentation_url%5C%22%3A%5C%22https%3A%2F%2Fdocs.github.com%2Frest%2Foverview%2Fresources-in-the-rest-api%23rate-limiting%5C%22%7D+ApiError+sentry%2Fshared_integrations%2Fclient%2Fbase.py+_request+False+sentry.tasks.derive_code_mappings.derive_code_mappings%22+organization.slug%3Aadeo-0b&sort=-message&statsPeriod=3d&yAxis=count%28%29) for this.
image

File: src/sentry/integrations/github/client.py [hunk 2]
```diff
@@ -118,7 +118,8 @@ def get_tree(self, repo_full_name: str, tree_sha: str) -> JSONData:
...
+                # Raise it so we can stop hammering the API
+                raise (e)
```
Comment (armenzg) [inline src/sentry/integrations/github/client.py:122]: We raise it inside of `get_tree` and `get_cache_repo_files` will also bubble it up. `get_trees_for_org` will handle it.

File: src/sentry/integrations/github/client.py [hunk 3]
```diff
@@ -138,22 +139,19 @@ def get_cached_repo_files(
         key = f"github:repo:{repo_full_name}:{'source-code' if only_source_code_files else 'all'}"
         repo_files: List[str] = cache.get(key, [])
         if not repo_files:
-            try:
```
Comment (armenzg) [inline src/sentry/integrations/github/client.py:141]: We just remove the try/except block and do a left shift. No code changes.

File: src/sentry/integrations/github/client.py [hunk 4]
```diff
@@ -188,6 +186,11 @@ def get_trees_for_org(
                 repo_files = self.get_cached_repo_files(full_name, branch)
                 trees[full_name] = RepoTree(Repo(full_name, branch), repo_files)
             logger.info("Using cached trees for Github org.", extra=extra)
+        except ApiError as e:
```
Comment (armenzg) [inline src/sentry/integrations/github/client.py:189]: If any of the calls to `get_cached_repo_files` produces an API rate limit handle it in this code block and do not try anymore.

Ideally we should block other tasks from trying before the API rate limit is reset but this is already an improvement and may prevent Github blocking the installation.

File: tests/sentry/integrations/github/test_integration.py [hunk 5]
```diff
@@ -120,7 +127,12 @@ def _stub_github(self):
             "archived": {
                 "archived": True,
             },
+            "xyz": {
+                "full_name": "Test-Organization/xyz",
+                "default_branch": "master",
+            },
         }
+        self.repositories = repositories
```
Comment (armenzg) [inline tests/sentry/integrations/github/test_integration.py:135]: This allows me to use `self.repositories` without refactoring everywhere where `repositories` is used.

File: tests/sentry/integrations/github/test_integration.py [hunk 6]
```diff
@@ -585,9 +597,43 @@ def test_github_prevent_install_until_pending_deletion_is_complete(self):
             organization=self.organization, integration=integration
         ).exists()
 
+    def get_installation_helper(self):
```
Comment (armenzg) [inline tests/sentry/integrations/github/test_integration.py:600]: We were sharing this logic between the two tests.

File: tests/sentry/integrations/github/test_integration.py [hunk 7]
```diff
@@ -585,9 +597,43 @@ def test_github_prevent_install_until_pending_deletion_is_complete(self):
...
+        integration = Integration.objects.get(provider=self.provider.key)
+        installation = integration.get_installation(self.organization.id)
+        return installation
+
+    @responses.activate
+    def test_get_trees_for_org_handles_rate_limit_reached(self):
+        """Test that we will not hit Github's API more than once when we reach the API rate limit"""
+        installation = self.get_installation_helper()
+        # This will force reaching for xyz before foo
+        cache.set(
```
Comment (armenzg) [inline tests/sentry/integrations/github/test_integration.py:613]: This is a trick that allows me not to do mocking to change the order of how we iterate through the repositories.

File: tests/sentry/integrations/github/test_integration.py [hunk 8]
```diff
@@ -585,9 +597,43 @@ def test_github_prevent_install_until_pending_deletion_is_complete(self):
...
+            "githubtrees:repositories:foo:Test-Organization",
+            [
+                {"full_name": "Test-Organization/bar", "default_branch": "main"},
+                {"full_name": "Test-Organization/xyz", "default_branch": "master"},
```
Comment (armenzg) [inline tests/sentry/integrations/github/test_integration.py:617]: This one will hit the API rate limit.

File: tests/sentry/integrations/github/test_integration.py [hunk 9]
```diff
@@ -52,6 +52,13 @@
         "status_code": 404,
         "body": {"message": "Not Found"},
     },
+    "xyz": {
+        "status_code": 403,
```
Comment (armenzg) [inline tests/sentry/integrations/github/test_integration.py:56]: You can see the status code in [the issue](https://sentry.io/organizations/sentry/discover/sentry:e24949f714a549078b15b600b9e23f4e/?field=message&field=organization.slug&interval=1h&name=ApiError%3A+%7B%22message%22%3A%22Not+Found%22%2C%22documentation_url%22%3A%22https%3A%2F%2Fdocs.github.com%2Frest%2Freference%2Fapps%23create-an...&project=1&query=message%3A%22An+unknown+error+has+ocurred.+%7B%5C%22message%5C%22%3A%5C%22API+rate+limit+exceeded+for+installation+ID+865116.%5C%22%2C%5C%22documentation_url%5C%22%3A%5C%22https%3A%2F%2Fdocs.github.com%2Frest%2Foverview%2Fresources-in-the-rest-api%23rate-limiting%5C%22%7D+ApiError+sentry%2Fshared_integrations%2Fclient%2Fbase.py+_request+False+sentry.tasks.derive_code_mappings.derive_code_mappings%22+organization.slug%3Aadeo-0b&sort=-message&statsPeriod=3d&yAxis=count%28%29):
image

**Review by snigdhas — APPROVED** [review]

**Pull request merged**